lib/node_modules/@stdlib/lapack/base/dlaswp/lib/dlaswp.js

```diff
@@ -0,0 +1,144 @@
+/**
+* @license Apache-2.0
+*
+* Copyright (c) 2024 The Stdlib Authors.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+'use strict';
+
+// MODULES //
+
+var isLayout = require( '@stdlib/blas/base/assert/is-layout' );
+var floor = require( '@stdlib/math/base/special/floor' );
+var dswap = require( '@stdlib/blas/base/dswap' ).ndarray;
+
+
+// MAIN //
+
+/**
+* Performs a series of row interchanges on the matrix `A` using pivot indices stored in `IPIV`.
+*
+* @param {string} order - storage layout
+* @param {PositiveInteger} N - number of columns of `A`
+* @param {Float64Array} A - matrix
+* @param {PositiveInteger} LDA - leading dimension of `A`
+* @param {PositiveInteger} k1 - index of first row to interchange
+* @param {PositiveInteger} k2 - index of last row to interchange
+* @param {Int32Array} IPIV - vector of pivot indices
+* @param {PositiveInteger} incX - increment between successive values of `IPIV`
+* @throws {TypeError} first argument must be a valid order
+* @returns {Float64Array} permuted matrix `A`
+*
+* @example
+* var Int32Array = require( '@stdlib/array/int32' );
+* var Float64Array = require( '@stdlib/array/float64' );
+*
+* var IPIV = new Int32Array( [ 2, 0, 1 ] );
+* var A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0, 5.0, 6.0 ] );
+*
+* dlaswp( 'column-major', 3, A, 2, 0, 2, IPIV, 1 );
+* // A => <Float64Array>[ 2.0, 3.0, 4.0, 5.0, 1.0, 6.0 ]
+*/
+function dlaswp( order, N, A, LDA, k1, k2, IPIV, incX ) {
```

Comment:
No `strideA` and no `incA`. Only `offsetA` and `offsetIPIV`.

I've updated `dspmv` to have an `offsetA` parameter. See https://github.com/stdlib-js/stdlib/commit/bc2355981ffa28082d8e23273887501bd3ba674a for an example.

Reply:
I do not think we need an `offsetIPIV` because `k1` is offset itself.

Reply:
Yeah, I suppose that could work; although, it is a little weird from a user POV, as `k1` needs to account for elements outside of a `IPIV` view. My initial impulse is that a separate parameter may be better here rather than overloading `k1`.

Reply:
what if then `offsetIPIV + k1 > k2`

Reply:
I think it will be confusing to provide `offsetIPIV` and also I am unable to think any usecase where we need both `k1` and `offsetIPIV`

Reply:
The use case is `ndarray`s, as the method's name should suggest. Every `ndarray` instance has an `offset`, which points to where the view starts in linear memory. Having an `.ndarray` method allows doing something like

```javascript
function dcopy( x, y ) {
	return blas_dcopy.ndarray( x.shape[ 0 ], x.data, x.strides[ 0 ], x.offset, y.data, y.strides[ 0 ], y.offset );
}
```

If we roll into `k1`, then would need to do something like

```javascript
dlaswp.ndarray( A.order, A.shape[ 0 ], A, A.offset, A.strides[ 0 ], k1 + IPIV.offset, k2 + IPIV.offset, IPIV.data, IPIV.strides[ 0 ] );
```

Note that now we need to explicitly update both `k1` and `k2`. We don't do something like this anywhere else, and, for other BLAS packages, if an interface takes a vector, the vector **always** has a separate offset parameter in an `.ndarray` method.

Reply:
Accordingly, you can make the argument that, given precedence elsewhere for `.ndarray` methods, **not** having a separate offset parameter is confusing, as it isn't obvious that one needs to explicitly shift **both** `k1` and `k2` when working with `ndarray` views.

Reply:
Right now we use pivot indices and replace `ith` row with `IPIV[i]`, once we introduce offset, we will replace `ith` row with `IPIV[i+offset]`, right? or `i+offset` row with `IPIV[i+offset]`. The latter is not correct.

Reply:
This means following diff should make it work:

```diff
diff --git a/lib/node_modules/@stdlib/lapack/base/dlaswp/lib/base.js b/lib/node_modules/@stdlib/lapack/base/dlaswp/lib/base.js
index acd1a0e0d7..dd641e7060 100644
--- a/lib/node_modules/@stdlib/lapack/base/dlaswp/lib/base.js
+++ b/lib/node_modules/@stdlib/lapack/base/dlaswp/lib/base.js
@@ -37,6 +37,7 @@ var dswap = require( '@stdlib/blas/base/dswap' ).ndarray;
 * @param {PositiveInteger} k1 - index of first row to interchange
 * @param {PositiveInteger} k2 - index of last row to interchange
 * @param {Int32Array} IPIV - vector of pivot indices
+* @param {PositiveInteger} offsetIPIV - index offset for `IPIV`
 * @param {PositiveInteger} strideIPIV - increment between successive values of `IPIV`
 * @returns {Float64Array} permuted matrix `A`
 *
@@ -47,10 +48,10 @@ var dswap = require( '@stdlib/blas/base/dswap' ).ndarray;
 * var IPIV = new Int32Array( [ 2, 0, 1 ] );
 * var A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0, 5.0, 6.0 ] );
 *
-* dlaswp( 'row-major', 3, A, 0, 2, 0, 2, IPIV, 1 );
+* dlaswp( 'row-major', 3, A, 0, 2, 0, 2, IPIV, 0, 1 );
 * // A => <Float64Array>[ 3.0, 4.0, 1.0, 2.0, 5.0, 6.0 ]
 */
-function dlaswp( order, N, A, offsetA, LDA, k1, k2, IPIV, strideIPIV ) {
+function dlaswp( order, N, A, offsetA, LDA, k1, k2, IPIV, offsetIPIV, strideIPIV ) {
        var inc;
        var ix0;
        var n32;
@@ -64,12 +65,12 @@ function dlaswp( order, N, A, offsetA, LDA, k1, k2, IPIV, strideIPIV ) {
        var k;
 
        if ( strideIPIV > 0 ) {
-               ix0 = k1;
+               ix0 = k1 + offsetIPIV;
                i1 = k1;
                i2 = k2;
                inc = 1;
        } else if ( strideIPIV < 0 ) {
-               ix0 = k1 + ( ( k1 - k2 ) * strideIPIV );
+               ix0 = k1 + ( ( k1 - k2 ) * strideIPIV ) + offsetIPIV;
                i1 = k2;
                i2 = k1;
                inc = -1;
```